core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1755,6 +1765,78 @@ class ReplicaManager(val config: KafkaConfig,
+    val traceLoggingEnabled = stateChangeLogger.isTraceEnabled
+    // Do we need this?
+    partitionStates.forKeyValue { (partition, partitionState) =>
+      if (traceLoggingEnabled)
+        stateChangeLogger.trace(s"Handling LeaderAndIsr request correlationId $correlationId from controller $controllerId " +
+          s"epoch $controllerEpoch starting the become-follower transition for partition ${partition.topicPartition} with leader " +
+          s"${partitionState.leader}")
+      responseMap.put(partition.topicPartition, Errors.NONE)
+    }
+
+    val partitionsToUpdateFollower: mutable.Set[Partition] = mutable.Set()
+    try {
+      partitionStates.forKeyValue { (partition, partitionState) =>
+        val newLeaderBrokerId = partitionState.leader
+          if (metadataCache.hasAliveBroker(newLeaderBrokerId)) {
+            // Only change partition state when the leader is available
+            partitionsToUpdateFollower += partition
+          } else {
+            // The leader broker should always be present in the metadata cache.
+            // If not, we should record the error message and abort the transition process for this partition
+            stateChangeLogger.error(s"Received LeaderAndIsrRequest with correlation id $correlationId from " +
+              s"controller $controllerId epoch $controllerEpoch for partition ${partition.topicPartition} " +
+              s"(last update controller epoch ${partitionState.controllerEpoch}) " +
+              s"but cannot become follower since the new leader $newLeaderBrokerId is unavailable.")
+          }
+      }
+
+      if (isShuttingDown.get()) {
+        if (traceLoggingEnabled) {
+          partitionsToUpdateFollower.foreach { partition =>
+            stateChangeLogger.trace(s"Skipped the update topic ID step of the become-follower state " +
+              s"change with correlation id $correlationId from controller $controllerId epoch $controllerEpoch for " +
+              s"partition ${partition.topicPartition} with leader ${partitionStates(partition).leader} " +
+              "since it is shutting down")
+          }
+        }
+      } else {
+        val partitionsToMakeFollowerWithLeaderAndOffset = partitionsToUpdateFollower.map { partition =>
+          val leaderNode = partition.leaderReplicaIdOpt.flatMap(leaderId => metadataCache.
+            getAliveBrokerNode(leaderId, config.interBrokerListenerName)).getOrElse(Node.noNode())
+          val leader = new BrokerEndPoint(leaderNode.id(), leaderNode.host(), leaderNode.port())
+          val log = partition.localLogOrException
+          val fetchOffset = initialFetchOffset(log)
+          partition.topicPartition -> InitialFetchState(topicIds(partition.topic), leader, partition.getLeaderEpoch, fetchOffset)
+        }.toMap
+
+        replicaFetcherManager.addFetcherForPartitions(partitionsToMakeFollowerWithLeaderAndOffset)
```

Comment:
I have been thinking about this. As this point, the fetcher is already running so we know that we only want to update the partition state with the newly assigned topic id. Creating the `InitialFetchState` seems a bit wasteful in this case, no? I wonder if it would not be better to add a new method to `AbstractFetcherManager` to basically update the partition states with the newly assigned topic ids instead. Is it something that you have already considered?

Reply:
I may have considered something like that. I think my biggest question is whether it is safe to update partition states in that way. I can take another look.

Reply:
There are a few ways to do this and I'll be interested to see which is the best. It is a little tricky to go from topic partition -> fetcher -> fetch state where we can update the topic ID. 

One way is to use `find` like `fetchState` does. Another is to look up the fetcher in the map like `addFetcherForPartitions` does. Finally, we could try to just add topic IDs to all the partitions in the fetcher by iterating through, but that might not be very efficient